prusti-viper/src/encoder/builtin_encoder.rs

```diff
@@ -133,6 +150,70 @@ impl BuiltinEncoder {
                     body: None,
                 }
             },
+            BuiltinFunctionKind::SliceLookupPure { slice_elem_ty, return_ty } => {
+                let elem_ty_name = if let vir::Type::TypedRef(ref ty) = slice_elem_ty { ty } else { unreachable!() };
+                let type_pred = format!("Slice${}", elem_ty_name);
+
+                let slice_len = self.encode_builtin_function_name(
+                    &BuiltinFunctionKind::SliceLen { slice_elem_ty }
+                );
+                let self_var = vir::LocalVar::new("self", vir::Type::TypedRef(type_pred.clone()));
+                let idx_var = vir_local!{ idx: Int };
+                let slice_len_call = vir::Expr::func_app(
+                    slice_len,
+                    vec![
+                        vir::Expr::local(self_var.clone()),
+                    ],
+                    vec![
+                        vir::LocalVar::new(String::from("self"), vir::Type::TypedRef(type_pred.clone())),
+                    ],
+                    vir::Type::Int,
+                    vir::Position::default(),
+                );
+
+                vir::Function {
+                    name: fn_name,
+                    formal_args: vec![
+                        self_var.clone(),
+                        idx_var.clone(),
+                    ],
+                    return_type: return_ty,
+                    pres: vec![
+                        // idx < Slice${ty}$len(self)
+                        vir!{ [vir::Expr::local(idx_var)] < [slice_len_call] },
+                        // acc(self, read$())
+                        vir::Expr::predicate_access_predicate(
+                            type_pred,
+                            vir::Expr::local(self_var),
+                            vir::PermAmount::Read,
+                        )
+                    ],
+                    posts: vec![],
+                    body: None,
+                }
+            },
+            BuiltinFunctionKind::SliceLen { slice_elem_ty } => {
+                let elem_ty_name = if let vir::Type::TypedRef(ty_name) = slice_elem_ty { ty_name } else { unreachable!() };
+                let slice_pred = format!("Slice${}", elem_ty_name);
+                let self_var = vir::LocalVar::new("self", vir::Type::TypedRef(slice_pred.clone()));
+
+                vir::Function {
+                    name: format!("Slice${}$len", elem_ty_name),
+                    formal_args: vec![
+                        self_var.clone(),
+                    ],
+                    return_type: vir::Type::Int,
+                    pres: vec![
+                        vir::Expr::predicate_access_predicate(
+                            slice_pred,
+                            vir::Expr::local(self_var),
+                            vir::PermAmount::Read,
+                        ),
+                    ],
+                    posts: vec![],
```

Comment:
I think for some programs you'll need to add `result >= 0` as postcondition. I think you can already test this with a program that contains `assert!(&[1, 2, 3].len() >= 0);` or something similar.

Reply:
that won't work yet, the `core::slice::<impl [i32]>::len` (that's how it's formatted in the MIR) function call doesn't know yet that its result is the same as `Slice$i32$len` -- i don't know how to `#[extern_spec]` that because a) not sure how to specify the module for that `<impl [i32]>`, just ignoring it doesn't work; and b) i have no way to refer to `Slice$i32$len` in `extern_spec` i.e. from rust.

i'll attempt to special-case it in `procedure_encoder` the way `panic`, `Box::new` et al. are.

Reply:
did the special-casing attempt, see last commit.